res/controllers/Numark-Mixtrack-Platinum-scripts.js

```diff
@@ -0,0 +1,1415 @@
+    });
+    this.pitch_bend_down = new components.Button({
+        inKey: 'rate_temp_down',
+        input: pitch_or_keylock,
+    });
+    this.pitch_bend_up.other = this.pitch_bend_down;
+    this.pitch_bend_down.other = this.pitch_bend_up;
+
+    var key_up_or_down = function (channel, control, value, status, group) {
+        this.is_pressed = this.isPress(channel, control, value, status);
+        if (this.is_pressed) {
+            if (this.other.is_pressed) {
+                // reset if both buttons are pressed
+                this.inSetValue(0.0);
+            }
+            else {
+                var old = this.inGetValue();
+                // the gui only allows 7 steps, we replicate that behavior here
+                if (Math.abs(old + this.direction) < 8.0) {
+                    this.inSetValue(Math.round(old + 1.0 * this.direction));
+                }
+            }
+        }
+    };
+    this.key_up = new components.Button({
+        inKey: 'pitch_adjust',
+        direction: 1,
+        input: key_up_or_down,
+    });
+    this.key_down = new components.Button({
+        inKey: 'pitch_adjust',
+        direction: -1,
+        input: key_up_or_down,
+    });
+    this.key_up.other = this.key_down;
+    this.key_down.other = this.key_up;
+
+    loop_base = function(midino, obj) {
+        return _.assign({
+            midi: [0x94 + midi_chan, midino],
+            on: 0x01,
+            sendShifted: true,
+            shiftChannel: true,
+            shiftOffset: -0x10,
+        }, obj);
+    };
+
+    this.alternate_manloop = new components.ComponentContainer({
+        loop_in: new components.HotcueButton(loop_base(0x38, {
+            number: 5,
+        })),
+        loop_out: new components.HotcueButton(loop_base(0x39, {
+            number: 6,
+        })),
+        loop_toggle: new components.HotcueButton(loop_base(0x32, {
+            number: 7,
+        })),
+        loop_halve: new components.HotcueButton(loop_base(0x34, {
+            number: 8,
+        })),
+        loop_double: new components.HotcueButton(loop_base(0x35, {
+            number: 8,
+        })),
+    });
+    this.normal_manloop = new components.ComponentContainer({
+        loop_in: new components.Button(loop_base(0x38, {
+            inKey: 'loop_in',
+            outKey: 'loop_start_position',
+            outValueScale: function (value) {
+                return (value != -1) ? this.on : this.off;
+            },
+        })),
+        loop_out: new components.Button(loop_base(0x39, {
+            inKey: 'loop_out',
+            outKey: 'loop_end_position',
+            outValueScale: function (value) {
+                return (value != -1) ? this.on : this.off;
+            },
+        })),
+        loop_toggle: new components.LoopToggleButton(loop_base(0x32, {})),
+        loop_halve: new components.Button(loop_base(0x34, {
+            key: 'loop_halve',
+            input: function(channel, control, value, status) {
```

Comment:
Why is this needed? Doesn't `loop_halve` with the default toggle Button behavior work? Likewise for `loop_double` below.

Reply:
It works, but it doesn't change the loop length for non-quantized loops. Setting `loop_scale` works for quantized and non-quantized loops. In fact the revision where `loop_halve` was changed to only work for quantized loops mentions this as a work around.

Reply:
Oh right, I forgot the context that this is in manual loop mode. Also, this mapping does not use `beatloop_toggle`, so updating `beatloop_size` doesn't really matter much. So I guess it is okay to leave it as is here.

Maybe we should introduce new ControlPushButtons for this use case instead of the unusual `loop_scale` Control. Figuring out how loop halve/double buttons should work when designing the current looping interface was very difficult and there may still be room for improvement. If you want to dig into the weeds of the discussion, it was in #1187.

Reply:
I added a note to the [wiki page](https://www.mixxx.org/wiki/doku.php/numark_mixtrack_platinum#manual_loop_mode) documenting this.